### ISSUE TITLE
Vector3 equals

### ISSUE THREAD

**Issue by michaelvogt** [google-ar/sceneform-android-sdk#342]
I might miss something trivial again, but shouldn't the result be true?

`boolean result = new Vector3(1, 1, 1).equals(new Vector3(1, 1, 1));`

**Comment by romainguy** [comment]
I believe it should be true yes :)

**Comment by dsternfeld7** [comment]
Vector3 does not override Object.equals, so it is just comparing the instances, not the content. You can use the static method Vector3.equals.

**Comment by romainguy** [comment]
@dsternfeld7 Any reason we don't override equals() here? It would make Kotlin better (== calls equals() in Kotlin).

**Comment by michaelvogt** [comment]
I see. I was only wondering as the decompiled code of Vector3 showed an equals override.

Yes, adding it would be a good thing. Appreciated for writing tests.

The same issue exists with Quaternion

**Comment by kstpr** [comment]
I've been dealing with geometric objects and floating point arithmetics in my code lately and had some similar experience when adding `equals`. Initially I used data classes in Kotlin which automatically generate `equals` and `hashCode` for you. But in the FPA world you don't do exact comparison, you check if two numbers are "close enough". For example a point `(x0, y0, z0)` belongs to a plane `ax + by + cz + d = 0` when `ax0 + by0 + cz0 + d` is not just exactly, but sufficiently close to `0`. If you look at the decompiled code of `Vector3.equals` it's using `MathHelper.almostEqualRelativeAndAbs` for comparison of the vector components. This defines a relation that is reflective, symmetric, but not necessarily transitive, hence not an equivalence relation. And equals needs to define an equivalence relation, else it can make a bit of a mess because it's used internally in Java data structures, see Effective Java for more detailed explanation. So I converted back all my data classes for geometric objects to regular classes and added an equivalent to the `equals` static method of `Vector3`. That's just my two cents on the issue.

**Comment by michaelvogt** [comment]
Well, I need to serialize/deserialize Vector3 and Quaternions. Needing the equal function in the Java sense for the serializer tests.

**Comment by dsternfeld7** [comment]
Thanks for letting us know all of your thoughts on this, we'll look into it.

**Comment by michaelvogt** [comment]
Well, I am not a 3D math expert, but I suspect that even the result of the static function isn't right. What I am missing here?

![image](https://user-images.githubusercontent.com/231274/46644382-fb94f900-cbba-11e8-9ea5-0ac092881151.png)

It seems like Unity implemented the equal function like this
```
if (this.x.Equals(vector3.x) && this.y.Equals(vector3.y))
        return this.z.Equals(vector3.z);
```

So I guess the result above should be true?

**Comment by kstpr** [comment]
@romainguy This is a solution that will work for @michaelvogt 's problem - that way he can serialize the vectors. And if he needs to do some actual geometry stuff he can use the companion method. I think of adopting that approach in my code too, thanks!

**Comment by michaelvogt** [comment]
Sorry for my earlier message. It obviously didn't transport what I wanted to say. Short version should have been, that I was wondering why this is only a feature request.

Obviously, handling of the objects for the 3D space is different from normal data handling. Still, these objects should behave like proper java objects for data handling. The reference to Effective Java is perfect.

**Comment by michaelvogt** [comment]
@romainguy anything that makes these objects behave correctly is fine for me :)

**Comment by tpsiaki** [comment]
Fixed in Sceneform 1.7 release.  Documentation here: https://developers.google.com/ar/reference/java/sceneform/reference/com/google/ar/sceneform/math/Vector3#equals(java.lang.Object)
